Add recipe for sharing state between Astro components (#3347)
withastro/docs: src/content/docs/en/recipes/sharing-state.mdx

---
title: Share State Between Astro Components

> description: Learn how to share state across Astro components — with Nano Stores.

Reviewer: ```suggestion
description: Learn how to share state across Astro components with Nano Stores.
```

i18nReady: true
type: recipe
---

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

:::tip

> Using Framework components? See [how to share state between Islands](/en/core-concepts/sharing-state/)!

Reviewer 2: Nit: `Framework components` -> `framework components` for consistency with other pages
```suggestion
Using framework components? See [how to share state between Islands](/en/core-concepts/sharing-state/)!
```

Author: good catch!

:::

When building an Astro website, you may need to share state across components. Astro recommends the use of [Nano Stores](https://github.com/nanostores/nanostores) for shared client storage.

## Recipe

1. Install Nano Stores:

  <PackageManagerTabs>
    <Fragment slot="npm">
    ```shell
    npm install nanostores
    ```
    </Fragment>
    <Fragment slot="pnpm">
    ```shell
    pnpm install nanostores
    ```
    </Fragment>
    <Fragment slot="yarn">
    ```shell
    yarn add nanostores
    ```
    </Fragment>
  </PackageManagerTabs>

2. Create a store. In this example, the store tracks whether a dialog is open or not:

  ```ts title="src/store.js"
  import { atom } from 'nanostores';

  export const isOpen = atom(false);
  ```

> 3. Import and use the store in a `<script>` tag in your components:

Reviewer: ```suggestion
3. Import and use the store in a `<script>` tag in the components that will share state.

The following `Button` and `Dialog` components each use the shared `isOpen` state to control whether a particular `<div>` is hidden or displayed:
```

  ```astro title="src/components/Button.astro"

>   <button id="openDialog">Open</button>

Reviewer 3: What do you think about adding the code fence?
```suggestion
---
---
  <button id="openDialog">Open</button>
```

Author: I removed it originally to reduce scroll-space, but actually it does look a lot more "Astro" with the code fences vs plain hrml. which is nice.

Author: added the code fence :D

  <script>
    import { isOpen } from '../store.js';
    

>     // Set the store to true when the button is clicked
>     function openDialog() {
>       isOpen.set(true);
>     }
> 
>     document.getElementById('openDialog').addEventListener('click', openDialog);

Reviewer: ```suggestion
    // Set the store to true when the button is clicked
    function openDialog() {
      isOpen.set(true);
    }

    // Add an event listener to the button
    document.getElementById('openDialog').addEventListener('click', openDialog);
```

  </script>
  ```

  ```astro title="src/components/Dialog.astro"

>   <div id="dialog" style="display: hidden">Hello world!</div>

Reviewer 3: ```suggestion
---
---
  <div id="dialog" style="display: hidden">Hello world!</div>
```

---
title: Share State Between Astro Components
description: Learn how to share state across Astro components — with Nano Stores.
i18nReady: true
type: recipe
---

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

> :::tip
>   Using Framework components? See [how to share state between Islands](/en/core-concepts/sharing-state/)!
> :::

Reviewer 4: ```suggestion
:::tip
Using Framework components? See [how to share state between Islands](/en/core-concepts/sharing-state/)!
:::
```

> When building an Astro website, you may need to share state across components. Astro recommend the use of [Nano Stores](https://github.com/nanostores/nanostores) for shared client storage.

Reviewer 4: ```suggestion
When building an Astro website, you may need to share state across components. Astro recommends the use of [Nano Stores](https://github.com/nanostores/nanostores) for shared client storage.
```

## Recipe

1. Install Nano Stores:

  <PackageManagerTabs>
    <Fragment slot="npm">
    ```shell
    npm install nanostores
    ```
    </Fragment>
    <Fragment slot="pnpm">
    ```shell
    pnpm install nanostores
    ```
    </Fragment>
    <Fragment slot="yarn">
    ```shell
    yarn add nanostores
    ```
    </Fragment>
  </PackageManagerTabs>

2. Create a store. In this example, the store tracks whether a dialog is open or not:

  ```ts title="src/store.js"
  import { atom } from 'nanostores';

  export const isOpen = atom(false);
  ```

> 3. Import and use the store in a [`script`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script) tag in your components:

Reviewer 4: Personal opinion; I don't think that we need to link the HTML docs here. If our readers are trying to implement stores, they probably know what HTML / JS is, thus are familiar with the `script` tag.

What do you think?

```suggestion
3. Import and use the store in a `<script>` tag in your components:
```

  ```astro title="src/components/Button.astro"

>   ---
>   ---

Reviewer 4: If we don't need the frontmatter, I would opt to totally remove it from the codesample.

```suggestion
```

  <button id="openDialog">Open</button>

  <script>
    import { isOpen } from '../store.js';
    
    // Set the store to true when the button is clicked
    function openDialog() {
      isOpen.set(true);
    }

    document.getElementById('openDialog').addEventListener('click', openDialog);
  </script>
  ```

  ```astro title="src/components/Dialog.astro"

>   ---
>   ---

Reviewer 4: see comment before
```suggestion
```

  <div id="dialog" style="display: hidden">Hello world!</div>

  <script>
    import { isOpen } from '../store.js';

    // Listen to changes in the store, and show/hide the dialog accordingly    
    isOpen.subscribe(open => {
      if (open) {
        document.getElementById('dialog').style.display = 'block';
      } else {
        document.getElementById('dialog').style.display = 'none';
      }
    })

    document.getElementById('openDialog').addEventListener('click', openDialog);
  </script>
  ```

## Resources

> - [Nano Stores - NPM](https://www.npmjs.com/package/nanostores)

Reviewer 4: ```suggestion
- [Nano Stores on NPM](https://www.npmjs.com/package/nanostores)
```

> - [Nano Stores - Vanilla JS docs](https://github.com/nanostores/nanostores#vanilla-js)

Reviewer 4: ```suggestion
- [Nano Stores documentation for Vanilla JS](https://github.com/nanostores/nanostores#vanilla-js)
```
